Fix4894 Using `Into` and `TryInto` instead of `From` and `TryFrom`

xiongmao86 opened the pull request:
If you added a new lint, here's a checklist for things that will be
checked during review or continuous integration.

- \[x] Followed [lint naming conventions][lint_naming]
- \[x] Added lint documentation

check again after every change:
- \[ ] `cargo test` passes locally
- \[ ] Executed `cargo dev update_lints`
- \[ ] Run `cargo dev fmt`
- \[ ] Added passing UI tests (including committed `.stderr` file)
---
Fixes #4894. Using `Into` and `TryInto` instead of using `From` and `TryFrom`.

*Please write a short comment explaining your change (or "none" for internal only changes)*
changelog:
Using `Into` and `TryInto` instead of using `From` and `TryFrom`.

rust-highfive commented:
r? @phansch

(rust-highfive has picked a reviewer for you, use r? to override)

xiongmao86 commented:
Oh, this is a draft for discussion, how can I disable `r?`

xiongmao86 commented on clippy_lints/src/into_and_try_into_instead_of_from_and_try_from.rs:
```diff
@@ -0,0 +1,59 @@
+                    if let Some(tr_ref) = wbp.bounds[0].trait_ref();
+                    if let Some(def_id) = tr_ref.trait_def_id();
+                    then {
+                        ["From", "TryFrom"].map(|e| {
+                            if cx.tcx.is_diagnostic_item(Symbol::intern(e), def_id) {
+                                span_lint(
+                                    cx,
+                                    INTO_AND_TRY_INTO_INSTEAD_OF_FROM_AND_TRY_FROM,
+                                    wp.span(),
+                                    "What is WHERE PREDICATE"
+                                );
+                            }
+                        });
```
@flip1995, `is_diagnostic_item` doesn't trigger the lint. I have put `span_lint` outside, then clippy spans the lint. What is the problem with my code?

xiongmao86 replied:
And modifying to `From<T>` and `TryFrom<T>` doesn't seem to work either.

flip1995 replied:
see my review comments

xiongmao86 commented:
r? @flip1995 
Can you take a look when you have time?

flip1995 commented on clippy_lints/src/into_and_try_into_instead_of_from_and_try_from.rs:
```diff
@@ -0,0 +1,59 @@
+                    if let Some(tr_ref) = wbp.bounds[0].trait_ref();
+                    if let Some(def_id) = tr_ref.trait_def_id();
+                    then {
+                        ["From", "TryFrom"].map(|e| {
```
I just saw, that only the `From` trait is a diagnostic item, not the `TryFrom` trait. I'll submit a patch for `try_from`. It may take a few days until this is usable in Clippy (next sync after the patch gets merged).

For the `From` case, try `cx.tcx.is_diagnostic_item(sym::from_trait, def_id)`.

flip1995 commented on clippy_lints/src/lib.rs:
```diff
@@ -1673,6 +1678,7 @@ pub fn register_plugins(store: &mut rustc_lint::LintStore, sess: &Session, conf:
         LintId::of(&functions::RESULT_UNIT_ERR),
         LintId::of(&if_let_some_result::IF_LET_SOME_RESULT),
         LintId::of(&inherent_to_string::INHERENT_TO_STRING),
+        LintId::of(&into_and_try_into_instead_of_from_and_try_from::INTO_AND_TRY_INTO_INSTEAD_OF_FROM_AND_TRY_FROM),
```
module/lint name is a little long. Just name it `into_instead_of_from`.

flip1995 commented:
@xiongmao86 You now should also have access to the diagnostic items for `try_from_trait` (and to both `Into` traits if needed). Implementation LGTM otherwise.

xiongmao86 commented:
@flip1995 Thanks for the guidance, I found toolchain update today and I `git pull upstream master`, and after `git rebase master fix4894`, I can't `git push -u origin fix4894`.

The error message is:
```
 ! [rejected]            fix4894 -> fix4894 (non-fast-forward)
error: failed to push some refs to 'https://github.com/xiongmao86/rust-clippy.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
and `git branch -v` says fix4894 is `[ahead 243, behind 6]`. 

What is the problem?

flip1995 commented:
When you rebase, you have to force push. Force pushing deletes your previous commits from the repository and you can't get them back (well you can, but the requires much more git-fu), so you should be careful.

If your rebase goes through without any conflicts (which I assume is the case here) force pushing is usually save after a rebase. So ``git push -u origin fix4894 --force-with-lease`` should solve your problem. (Always use `--force-with-lease` instead of `-f` or `--force`, it gives you another safeguard).

xiongmao86 commented:
@flip1995 , May be I mistake `-f` for `-u`. The git docs says "-u" means adding upstream reference. I don't really understand, what is it used for in this case?

xiongmao86 commented:
@flip1995, I was thinking maybe I expressed this item in the wrong way, maybe the name should be `FromInsteadOfInto`, so I can said "allow FromInsteadOfInto"? Is it right? I am not sure?

xiongmao86 commented:
And about the `-u` option, I read the `branch.<name>.merge` part of `git config`, there are a lot concept to make clear, so I don't know how to ask properly.

The description is:
```
-u
--set-upstream
For every branch that is up to date or successfully pushed, add upstream (tracking) reference, used by argument-less git-pull(1) and other commands. For more information, see branch.<name>.merge in git-config(1).
```

And I see now the point is to understand the concept of `upstream(tracking) reference`, could you tell me what is it, or could you show me the way to make sense of it?

flip1995 commented:
> The git docs says "-u" means adding upstream reference. I don't really understand, what is it used for in this case?

So `git push` needs an address to know where to push the commits. So `git push` is actually just a shortcut for `git push <remote> <branch>` (`git push origin fix4894`). To be able to use this shortcut you first have to tell git about the address it should use for this shortcut. You do this with `-u` aka `--set-upstream`. You know if you have set this shortcut if you see the line

```
Your branch is up-to-date with 'origin/master'.
```

when running `git status`.

> I was thinking maybe I expressed this item in the wrong way, maybe the name should be `FromInsteadOfInto`, so I can said "allow FromInsteadOfInto"? Is it right? I am not sure?

I'm not really sure what you mean. The name of the struct does not really matter for the lint. The lint name comes from the all-caps name in the `declare_clippy_lint!` macro.

flip1995 commented:
> And I see now the point is to understand the concept of `upstream(tracking) reference`, could you tell me what is it, or could you show me the way to make sense of it?

You can understand `upstream` as the online repository vs your local checkout on your computer.

xiongmao86 commented:
@flip1995, thanks for your explanation, I see what -u is using now.

About the lint name, I have not been clear. I used to write lint name `IntoInsteadOfFrom` `Into` in front and `From` at last, but I think `FromInsteadOfInto`, `From` in front, and `Into` at last would be more appropriate.

And to read`#[allow(from_instead_of_into)]` as "allow using from trait instead of using into trait".  If the lint wouldn't be triggered if we `allow` it. Is this understanding correct?

xiongmao86 commented:
I think I may have made a mistake when I think of naming the lint. I want to make sure the naming is right.

flip1995 commented:
Ah good catch. Yeah `from_instead_of_into` is the better name :+1:

xiongmao86 commented:
It's Spring Festival Eve today, @flip1995, thank you for helping me a lot last year, may you be healthy and happy and good lucky in the new year.

xiongmao86 commented:
@flip1995, I am trying to make a suggestion, how can I get the string of `T` of `From<T>` so that I can put it into `format!`? I have traced along `WhereClause` and did not find a ty. I guess probably I should use an id, `DefId` or any other `Id`, but I don't know how. I have search `Generic` and other word in `utils` mod, but I didn't find  anything fits my expectation.

flip1995 commented:
This is really deep in the where clause. Finally it is the `args` field of the last `PathSegment` of `From<T>`. You can get to this with:

[`WhereClause::predicates`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/struct.WhereClause.html#structfield.predicates) > `WherePredicate::BoundPredicate(WhereBoundPredicate)` > `WhereBoundPredicate::bounds` > `GenericBound::Trait(PolyTraitRef)` > `PolyTraitRef::trait_ref` > `TraitRef::path` > `Path::segments.last()` > [`PathSegment::args`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.PathSegment.html#structfield.args)

I linked the start and the finish for you. You can click through the documentation yourself ;)

xiongmao86 commented:
Thank you, @flip1995, I stopped when I reach path segment, I have read the parser before but I forgot there are generic args in path, I should have dig deeper.

xiongmao86 commented:
@flip1995, I think the `T: From<U>` in `needless_question_mark` is irrelevant, do you agree?

For fixed test to compile after changing from `TryFrom` to `TryInto`, it's necessary  to remove `import ...:TryFrom` and add `import ...:TryInto` to the fixed code, I don't think this is subject to change of the lint, do you agree? How can I make the test to pass?

flip1995 commented:
> I think the `T: From<U>` in `needless_question_mark` is irrelevant, do you agree?

Yes just allow the lint there.



> I don't think this is subject to change of the lint, do you agree? How can I make the test to pass?

Yes, this should not be included in the lint suggestion. You can make the test pass by also importing `TryInto`/`Into` and then add a `#![allow(unused_imports)]` to the top of the test file.

flip1995 commented on clippy_lints/src/from_instead_of_into.rs:
```diff
@@ -0,0 +1,82 @@
+                                "TryInto trait is preferable than TryFrom as a generic bound",
+                                "try",
+                                sugg,
+                                Applicability::MachineApplicable
```
This should be 
```suggestion
                                Applicability::MaybeIncorrect
```

because, `TryInto` may not be imported.

flip1995 commented on clippy_lints/src/from_instead_of_into.rs:
```diff
@@ -0,0 +1,82 @@
+        match wp {
+            WherePredicate::BoundPredicate(wbp) => {
+                if_chain! {
+                    if let Some(tr_ref) = wbp.bounds[0].trait_ref();
```
The ``From`` bound does not have to be the first bound in the predicate. So you could use `wbp.bounds.iter().find(|b| cond(b))` here instead.

flip1995 commented on tests/ui/from_instead_of_into.rs:
```diff
@@ -0,0 +1,16 @@
+
+fn foo<T>(a: T)
+where
+    u32: From<T>,
```
Please also add tests where `From` is not the only predicate in the bound. (Also put it in different positions)

flip1995 reviewed (commented):
LGTM, except that you assume that `From` is the first and only bound in the predicate. This needs testing and probably a fix.

xiongmao86 commented on clippy_lints/src/from_instead_of_into.rs:
```diff
@@ -74,7 +74,17 @@ impl LateLintPass<'tcx> for FromInsteadOfInto {
+                            if wbp.bounds.len() == 1 {
+                                sugg = extracted_where_predicate;
+                            } else {
+                                let bounds: &[&GenericBound<'_>] = wbp.bounds.iter().filter(|b| ptr::eq(*b, target_bound)).collect();
+                                let bounds = bounds.map(|b| snippet(cx, b.span(), ".."));
+                                let first = bounds[0].to_string();
+                                let bounds_str = bounds[1..].fold(first, |s, &e| s + " + " + snippet(cx, e.span(), ".."));
```
@flip1995 , I couldn't get this compile, what should I do?

flip1995 replied:
I don't think you have to make this that complicated. You don't need to include other bounds or the bounded type variable in the suggestion. Just use the span of the bound (`From`/`TryFrom`) and suggest to replace it with `Into`/`TryInto`

xiongmao86 replied:
@flip1995, I can't get what you mean, so if there is an `u32: Copy + From<T> + Clone` binding, if I extract `From<T>` out, what I think is to change it into `T: Into<u32>, u32: Copy + Clone`, how do you think I should handle this?

flip1995 replied:
Oh right, you have to swap the generic parameter... Yeah, that makes things more complicated. I have to test how to best do this myself.

xiongmao86 replied:
Thank you very much, @flip1995, I really don't know how to proceed.

xiongmao86 replied:
Thank you, @flip1995, after reading your suggestion, I did some research on `Span` and on `Iterator`, I think I can do this. 

But there are some part I still have doubt, when you refer to 2 part suggestion you are not referring to multi span suggestion, do I understand correctly?

flip1995 commented on clippy_lints/src/from_instead_of_into.rs:
```diff
@@ -0,0 +1,104 @@
+                    if let Some(generic_arg) = last_seg.args().args.first();
+                    if wbp.bounds.len() == 1;
+                    then {
+                        let bounded_ty = snippet(cx, wbp.bounded_ty.span, "..");
+                        let generic_arg_of_from_or_try_from = snippet(cx, generic_arg.span(), "..");
```
You should use `snippet_opt` here and put those in the `if_chain!`

flip1995 commented on clippy_lints/src/from_instead_of_into.rs:
```diff
@@ -0,0 +1,104 @@
+                        let bounded_ty = snippet(cx, wbp.bounded_ty.span, "..");
+                        let generic_arg_of_from_or_try_from = snippet(cx, generic_arg.span(), "..");
+
+                        let mut replace_trait_name = "";
+                        let mut target_trait_name = "";
```
```suggestion
                        let replace_trait_name;
                        let target_trait_name;
```

xiongmao86 commented on tests/ui/from_instead_of_into.rs:
```diff
@@ -0,0 +1,29 @@
+
+fn bar1<T>(a: T)
+where
+    u32: Copy + TryFrom<T> + Clone,
```
Doubts: I have added two multiply generic bound predicate here, do I need more predicate?

xiongmao86 replied:
How should I add more generic bounds?

Lengthen the origin bounds and put `From<T>` at the first, middle, and last should be enough?

xiongmao86 commented on tests/ui/from_instead_of_into.rs:
```diff
@@ -0,0 +1,29 @@
+}
+
+fn foo1<T>(a: T)
+where u32: Copy + Clone + From<T>,
```
Another predicate here. Is it not enough here?

I can't imagine a more complicated predicate that is also meaningful here. My understanding of generic bounds is rather shallow, I didn't know how to deepen my understanding, I have tried to search `rust generic bound` but didn't get progress here. How do you suggest I can learn about it? How did you learn about generic bounds in the past?

xiongmao86 commented:
@flip1995, I have implemented your suggestion. Although I have added `import std::convert::TryInto;` to test file, the added test still failed in the workspace test. What is the problem?

bors commented:
:umbrella: The latest upstream changes (presumably #6717) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

flip1995 commented on clippy_lints/src/from_instead_of_into.rs:
```diff
@@ -0,0 +1,140 @@
+
+                        let extracted_where_predicate = format!("{}: {}<{}>", generic_arg_of_from_or_try_from, replace_trait_name, bounded_ty);
+                        let sugg;
+                        if bounds.len() == 1 {
+                            sugg = extracted_where_predicate;
+                        } else {
+                            let bounds = get_reduced_bounds_str(cx, position, bounds);
+                            sugg = format!("{}, {}: {}", extracted_where_predicate, bounded_ty, bounds);
+                        }
+                        span_lint_and_sugg(
+                            cx,
+                            FROM_INSTEAD_OF_INTO,
+                            wp.span(),
+                            &message,
+                            "try",
+                            sugg,
+                            Applicability::MaybeIncorrect
+                        );
```
So instead of trying to rebuild the whole predicate string without that one bound, it would be easier so just get the span of the bound and suggest to remove it. 

> But there are some part I still have doubt, when you refer to 2 part suggestion you are not referring to multi span suggestion, do I understand correctly?

No. You should use `span_lint_and_then` here instead and add 2 suggestions if there is more than one bound. Just grep for `span_lint_and_then` in the repo and you'll find how to do this.

xiongmao86 commented on clippy_lints/src/from_instead_of_into.rs:
```diff
@@ -0,0 +1,119 @@
+                    }
+                    let removed_span = target_bound.span().with_lo(low);
+
+                    span_lint_and_then(
+                        cx,
+                        FROM_INSTEAD_OF_INTO,
+                        wp.span(),
+                        &message,
+                        |diag| {
+                            diag.span_suggestion(
+                                removed_span,
+                                &format!("remove {} bound", target_trait_name),
+                                "".to_string(),
+                                Applicability::MaybeIncorrect,
+                            );
+
+                            let sugg;
+                            if bounds.len() == 1 {
+                                sugg = switched_predicate;
+                            } else {
+                                sugg = format!(", {}", switched_predicate);
+                            }
+                            diag.span_suggestion(
+                                wp.span().with_lo(wp.span().hi()),
+                                "Add this bound predicate",
+                                sugg,
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
+                    );
```
When I handle suggestion with this two `span_suggestion` call the diagnostic output is strange.  Remove suggestion is weird because nothing show up in the span only a `,` left. replacing suggestion is weird because the old span is used. Can you take a look? @flip1995 .

xiongmao86 replied:
Could you tell me why you suggest using multi suggestions? Do you intent to suggest using `multispan_sugg`? If yes, why do you choose it?

bors commented:
:umbrella: The latest upstream changes (presumably #6573) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

xiongmao86 commented on tests/ui/from_instead_of_into.stderr:
```diff
@@ -0,0 +1,78 @@
+   |     ^^^^^^^^^^^^
+   |
+   = note: `-D clippy::from-instead-of-into` implied by `-D warnings`
+help: remove From bound
+   |
+LL |     ,
+   |    --
```
This is probably not the best diagnostic message, `From<T>` is removed and only comma left.

flip1995 replied:
Yeah, I don't even know if that still compiles. I think you have to track how many bounds there are and tweak the suggestion depending on this.

xiongmao86 commented on tests/ui/from_instead_of_into.stderr:
```diff
@@ -0,0 +1,78 @@
+   |
+LL |     ,
+   |    --
+help: Add this bound predicate
+   |
+LL |     u32: From<T>T: Into<u32>,
+   |                 ^^^^^^^^^^^^
+
```
I made a mistake here. Didn't notice until now. I'll fix this.

xiongmao86 commented on tests/ui/from_instead_of_into.stderr:
```diff
@@ -0,0 +1,78 @@
+   |                     --
+help: Add this bound predicate
+   |
+LL |     u32: Copy + Clone + From<T>, T: Into<u32>,
```
I think this probably should be `u32: Copy + Clone, T: Into<u32>`, but I don't know how to remove `From` showing in this span.

flip1995 replied:
You have to get the span of the `From<T>` bound and extend it to the end of the span of the `Clone` bound with `span.with_lo(clone_bound_span.hi())` And as the suggestion you can just use `String::new()`, which will automatically suggest removing it.

xiongmao86 commented:
@flip1995, are you suggesting using `multispan_sugg`?

bors commented:
:umbrella: The latest upstream changes (presumably #6787) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

flip1995 commented:
You can use `multispan_sugg` or just add 2 suggestions, by calling `diag.span_suggestion` 2 times with different spans and suggestions. I would first try the latter. If that doesn't work or looks weird, I would try `mutlispan_sugg`.

flip1995 commented:
Sorry for being absent for so long. I currently can't really keep up with my reviews...

xiongmao86 commented:
Sorry, @flip1995, seems github.com is ban again at my location. I am trying to get to github.com(read prs, push commits etc.) in a secure way.

xiongmao86 commented:
> Sorry for being absent for so long. I currently can't really keep up with my reviews...

Thank you very much for your help. Please take your time.

giraffate commented:
ping from triage @xiongmao86. Can you have any update on this?

xiongmao86 commented:
@giraffate, I recently don't have time for this, if you want to take over, please go ahead.

giraffate commented:
ping from triage @xiongmao86. According to the triage procedure, I'm closing this because 2 weeks have passed with no activity.
